### PR TITLE
 discover k8s Gateway resource as network gateway

### PR DESCRIPTION
**Pull request by stevenctl**
Adds the ability to read Gateway API gateways to discover multi-network gateways. 

* Must have the topology.istio.io/label object
* Will not discover Gateways with class "istio-remote" from anything but the local cluster/config cluster. 
* Guarded by env variable

I do not want to add a release note or highlight this until this is more fleshed out.
It needs more work to support ambient, and I want to take another look at how we deal with:

* Overlap with the gateway defined by the Service. Right now if the address/port/network are the same, we replace it. We may want to override some piece of what's on the service to support things like #45903
* Multiple "auto-passthrough" listeners/ports. Should that expand to multiple gateways? This throws off the weight. Uncommon use case though.

### CODE REVIEW

File: pilot/pkg/autoregistration/controller.go [hunk 1]
```diff
@@ -662,13 +662,6 @@ func workloadEntryFromGroup(name string, proxy *model.Proxy, groupCfg *config.Co
 			Namespace:        proxy.Metadata.Namespace,
 			Labels:           entry.Labels,
 			Annotations:      annotations,
-			OwnerReferences: []metav1.OwnerReference{{
```
Comment (hzxuzhonghu) [inline pilot/pkg/autoregistration/controller.go:665]: I think this is unrelated change

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 2]
```diff
@@ -313,6 +330,79 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
...
+	autoPassthrough := func(l v1beta1.Listener) bool {
+		return kube.IsAutoPassthrough(gw.GetLabels(), l)
+	}
+
+	base := model.NetworkGateway{
+		Network: network.ID(gw.GetLabels()[label.TopologyNetwork.Name]),
+		Cluster: n.clusterID,
+	}
+	newGateways := model.NetworkGatewaySet{}
+	for _, addr := range gw.Spec.Addresses {
```
[the remark this answers is not in the document]
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:377]: I think this would be incredibly uncommon.

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    port: 15012
+    protocol: TLS
+    tls:
+      mode: Passthrough
+  - name: istiod-webhook
+    port: 15017
+    protocol: TLS
+    tls:
+      mode: Passthrough
+  - name: cross-network
```
Comment (costinm) [inline pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:22]: Why not 15443 ?

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    networking.istio.io/gatewayPort: 35443
+spec:
+  gatewayClassName: istio
+  listeners:
+  - name: istiod-grpc
+    port: 15012
+    protocol: TLS
+    tls:
+      mode: Passthrough
+  - name: istiod-webhook
```
Comment (costinm) [inline pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:17]: Do we still need to pass 15017 ? If I remember correctly, it was used for 'central istiod', but doesn't make sense for the 'istiod per cluster' model. I would either drop it or have a separate Gateway test for exposing central Istiod. 

I guess the same goes for 15012 - a remote gateway doesn't need to expose the remote Istiod. 

I know this is a test - but will confuse readers and may accidentally be used this way.

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: Gateway
+metadata:
+  name: eastwestgateway
+  namespace: istio-system
+  labels:
+    topology.istio.io/network: "network-1"
```
Comment (costinm) [inline pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:7]: Istiod in the cluster is supposed to know the network id it is operating on and to add it to all pods - i.e. the controller or template should already handle this automatically.  

No need for the user to customize a Gateaway per network and remember to copy the same name used at istiod install. 

How about having a new label to indicate this gateway is designated for E/W ? 

I guess you are setting it here for convenience in the code ( to get the network from watching the gateway ), but it becomes a bit harder for the user. 

May not be a huge deal and we may fix it when we move to hbone...

File: pilot/pkg/serviceregistry/kube/conversion.go [hunk 6]
```diff
@@ -233,3 +236,29 @@ func PodTLSMode(pod *corev1.Pod) string {
 	}
 	return model.GetTLSModeFromEndpointLabels(pod.Labels)
 }
+
+// IsAutoPassthrough determines if a listener should use auto passthrough mode. This is used for
+// multi-network. In the Istio API, this is an explicit tls.Mode. However, this mode is not part of
```
Comment (costinm) [inline pilot/pkg/serviceregistry/kube/conversion.go:241]: It is used for multi-network to automatically expose all mesh services on the gateway.

Which is a security risk and a major problem in the multi-network design, and quite inconsistent with  the new Gateway API.

I am quite uncomfortable to this - I was hoping we either start using HBone for multi-network ( so there is at least some authentication ) or explicit opt in for services to be exposed in the E/W.

Note that Gateway API does NOT send traffic even to multiple clusters in the same network - there is an explicit opt-on via backendRef of ServiceImport kind.

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 7]
```diff
@@ -313,6 +330,79 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
 	return nil
 }
 
+func (n *networkManager) watchGatewayResources(c *Controller, stop <-chan struct{}) {
+	n.gatewayResourceClient = kclient.NewDelayedInformer(c.client, gvr.KubernetesGateway, kubetypes.StandardInformer, kubetypes.Filter{
+		LabelSelector: label.TopologyNetwork.Name,
+		FieldSelector: "spec.gatewayClassName=istio",
```
Comment (costinm) [inline pilot/pkg/serviceregistry/kube/controller/network.go:336]: I don't think we should or need to add this restriction. 

A user may use other Gateway classes - as long as they provide a TLS pass through, we should not 
restrict it to istio class.

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 8]
```diff
@@ -313,6 +330,79 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
...
+
+	previousGateways := n.gatewaysFromResource[gw.UID]
+
+	if event == model.EventDelete {
+		gatewaysChanged = len(previousGateways) > 0
+		delete(n.gatewaysFromResource, gw.UID)
+		return nil
+	}
+
+	autoPassthrough := func(l v1beta1.Listener) bool {
```
Comment (costinm) [inline pilot/pkg/serviceregistry/kube/controller/network.go:368]: Can you add a comment somewhere on where is this used ? Is it to configure the istio gateway to enable auto-passthrough in the backend ? Or to detect if the remote gateway has the 15443 port and can be used to route ? 

If the later - I don't think there is a strict requirement that any E/W gateway is forwarding EVERYTHING. Even in current implementation an admin may use NetworkPolicy and other controls to limit the pods that are publicly exposed ( because our E/W remains a pretty serious risk ). 
Having port 15443 with a certain name tells us this is a gateway that may forward mTLS to backends in a different network.

File: pilot/pkg/config/kube/gateway/testdata/eastwest-namedport.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    port: 15012
+    protocol: TLS
+    tls:
+      mode: Passthrough
+  - name: istiod-webhook
+    port: 15017
+    protocol: TLS
+    tls:
+      mode: Passthrough
+  - name: tls
```
Comment (costinm) [inline pilot/pkg/config/kube/gateway/testdata/eastwest-namedport.yaml:21]: Not sure I see the difference with the previous test ( labelport ). 

Is this intended as a negative test ? I think the name of the 15443 port should be very explicit and used in the check, not 'tls'

File: pilot/pkg/serviceregistry/kube/conversion.go [hunk 10]
```diff
@@ -233,3 +236,29 @@ func PodTLSMode(pod *corev1.Pod) string {
...
+// a multi-network gateway, so we will use this as a signal.
+// A user who wishes to expose multi-network connectivity should create a listener named "tls"
+// with TLS.Mode Passthrough.
+// For some backwards compatibility, we assume any listener with TLS specified and a port matching
+// 15443 (or the label-override for gateway port) is auto-passtrough as well.
+func IsAutoPassthrough(gwLabels map[string]string, l v1beta1.Listener) bool {
+	if l.TLS == nil {
+		return false
+	}
+	if l.Name == "tls" && l.TLS.Mode != nil && *l.TLS.Mode == v1beta1.TLSModePassthrough {
```
Comment (costinm) [inline pilot/pkg/serviceregistry/kube/conversion.go:252]: I would not name the port 'tls' - far too generic name. 

'tls-passthrough' or 'tls-multi-network' or something very specific.

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml (hunk 4 above)
Comment (stevenctl) [reply pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:17]: You're probably right. This is mostly copypaste. I'll remove it.

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml (hunk 5 above)
Comment (stevenctl) [reply pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:7]: The intent is that I am writing a Gateway resource to represent a _remote_ gateway. Similar to the WorkloadEntry proposal, this allows discovery of remote resources without watching a remote API server.

File: pilot/pkg/serviceregistry/kube/conversion.go (hunk 6 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/conversion.go:241]: Don't we have opt-in via `servers.hosts`? Not as nice as  an attached backendRef, but HTTPRoute won't make sense for TCP workloads.

File: pilot/pkg/serviceregistry/kube/conversion.go (hunk 10 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/conversion.go:252]: Agree.

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 11]
```diff
@@ -313,6 +333,83 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
 	return nil
 }
 
+func (n *networkManager) watchGatewayResources(c *Controller, stop <-chan struct{}) {
+	n.gatewayResourceClient = kclient.NewDelayedInformer(c.client, gvr.KubernetesGateway, kubetypes.StandardInformer, kubetypes.Filter{
+		LabelSelector: label.TopologyNetwork.Name,
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/network.go:338]: This will open a new watch, when we already have one. Should prefer to just filter on the clientside IMO

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 12]
```diff
@@ -577,16 +577,15 @@ func (c *Controller) HasSynced() bool {
...
-	}
-	return true
+	return c.namespaces.HasSynced() &&
+		c.services.HasSynced() &&
+		c.endpoints.slices.HasSynced() &&
+		c.pods.pods.HasSynced() &&
+		c.nodes.HasSynced() &&
+		c.imports.HasSynced() &&
+		c.exports.HasSynced() &&
+		c.client.CrdWatcher().HasSynced() &&
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:587]: I don't think we need to check crdwatcher

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 13]
```diff
@@ -614,6 +613,12 @@ func (c *Controller) Run(stop <-chan struct{}) {
 	go c.imports.Run(stop)
 	go c.exports.Run(stop)
 
+	if c.client.CrdWatcher().KnownOrCallback(gvr.KubernetesGateway, func(stop <-chan struct{}) {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:616]: we can just always start it, and it will do nothing if the CRD does not exist

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 14]
```diff
@@ -313,6 +333,83 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
...
+	})
+	registerHandlers[controllers.Object](c, n.gatewayResourceClient, "Gateways", n.handleGatewayResource, nil)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/network.go:340]:
```suggestion
	registerHandlers[*v1beta1.Gateway(c, n.gatewayResourceClient, "Gateways", n.handleGatewayResource, nil)
```

then you don't need to cast

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    networking.istio.io/gatewayPort: 35443
+spec:
+  gatewayClassName: istio
```
Comment (kdorosh) [inline pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:10]: for ambient we're going to need two different ways to represent SANs for verification during double HBONE (SAN verification for the gateway _and_ destination). the dest workload / svc in remote cluster's SAN can be derived from the service account on the workload entry.

another place we could leverage in theory is `subjectAltNames` on the service entry (if we took that approach as opposed to gateway API here). i am also open to using gateway API here so long as we leave a way to represent the expected SAN of the network gateway

i understand this is not a blocker outside of ambient, but i want to make sure what we go with here will also work for ambient

File: pilot/pkg/config/kube/gateway/testdata/eastwest-namedport.yaml (hunk 9 above)
Comment (stevenctl) [reply pilot/pkg/config/kube/gateway/testdata/eastwest-namedport.yaml]: Tests we can infer from name `tls-passthrough`

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 12 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:587]: Didn't realize the delayedClient baked a HasSynced check on CrdWatcher. `kclient` is pretty nice :)

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 14 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:340]: AFAICT, I can't convert `kclient.Untyped` to an `Informer[*v1beta1.Gateway]` (or anything that changes the generic)

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 16]
```diff
@@ -604,6 +601,8 @@ func (c *Controller) Run(stop <-chan struct{}) {
 	go c.imports.Run(stop)
 	go c.exports.Run(stop)
 
+	c.networkManager.watchGatewayResources(c, stop)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:604]: donot we only need to watch gateway for config cluster?

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 17]
```diff
@@ -15,28 +15,38 @@
...
 type networkManager struct {
 	sync.RWMutex
 	// CIDR ranger based on path-compressed prefix trie
-	ranger              cidranger.Ranger
-	clusterID           cluster.ID
-	meshNetworksWatcher mesh.NetworksWatcher
+	ranger    cidranger.Ranger
+	clusterID cluster.ID
+
+	gatewayResourceClient kclient.Untyped
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/network.go:48]: It does not seem quite right, we only support watching gateway/other CRD in config cluster.

The proper change should be add a CB handler to CRD client

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 18]
```diff
@@ -313,6 +333,86 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
...
+		return nil
+	}
+
+	if nw := gw.GetLabels()[label.TopologyNetwork.Name]; nw == "" {
+		return nil
+	}
+
+	// we don't want to discover gateways with class "istio-remote" from outside cluster's API servers.
+	// this will only pass for the "config cluster" or "local cluster"
+	if !n.discoverRemoteGatewayResources && gw.Spec.GatewayClassName == "istio-remote" {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/network.go:357]: I can not understand this well, though it has a comment.

Is it doing some handling with `istio-remote`  for non config cluster?

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 19]
```diff
@@ -313,6 +333,86 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
...
+
+	base := model.NetworkGateway{
+		Network: network.ID(gw.GetLabels()[label.TopologyNetwork.Name]),
+		Cluster: n.clusterID,
+	}
+	newGateways := model.NetworkGatewaySet{}
+	for _, addr := range gw.Spec.Addresses {
+		for _, l := range slices.Filter(gw.Spec.Listeners, autoPassthrough) {
+			networkGateway := base
+			networkGateway.Addr = addr.Value
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/network.go:390]: The `addr` here can be not ip, it can be of type Hostname and NamedAddress as well

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 20]
```diff
@@ -313,6 +333,86 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
...
+			networkGateway.Port = uint32(l.Port)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/network.go:391]: for what case can it have multiple ports for network gateway

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 19 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:390]: I think we should disallow `NamedAddress` but hostname should definitely be allowed.

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 18 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:357]: I need to provide a better explanation for `istio-remote`. 

`Gateway` with `istio-remote`: only discover this from the config cluster. This is a way to reference a gateway that lives in a place that we are not watching the API server for. 

`Gateway` with other classes: discover these anywhere.

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 17 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:48]: > other CRD in config cluster

Not completely true, we do support watching workloadentry.

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 16 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:604]: We need to watch all clusters to discover their multi-network gateways. 
If the other clusters don't have the CRD, this won't do anything.

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 17 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:48]: > The proper change should be add a CB handler to CRD client

Not sure what you mean. According to John, the way we initialize this only takes effect with the CRD present and is robust to the CRDs being added later.

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml (hunk 15 above)
Comment (stevenctl) [reply pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:10]: "downstream" TLS is part of the spec: https://gateway-api.sigs.k8s.io/guides/tls/?h=tls

We can define an HBONE listener for the east-west gateway to be `proto: HTTP, tls: TERMINATE`. 

The only available kind for `certificateReferences` is `Secret`. This doesn't quite make sense for us. 

There is an [extensions](https://gateway-api.sigs.k8s.io/guides/tls/?h=tls#extensions) field we can use. 

> Gateway TLS configurations provides an options map to add additional TLS settings for implementation-specific features.

Comment (stevenctl) [reply pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:10]: Ideally we merge this PR first, and then add the capability for discovering [trusted gateway](https://docs.google.com/document/d/1MCRZWvlbZS6DLn5N4pD9ww2O-fL7jzi9-l0cWbqH1kk/edit#heading=h.4y7xkspbqkjh) identities along with the rest of the trusted gateway impl. 

Confident that there is a place in both our old Service API as well as somewhere we can at least (temporarily) use in GAMMA.

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 19 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/network.go:390]: prevent by admission webhook? Who will implement that

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 16 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:604]: A little bit different from what istio does today, we do not support any other CR from remote clusters. So want to hear more voices about this

@howardjohn @ramaraochavali WDYT

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 17 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/network.go:48]: same as above comment

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 20 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/network.go:391]: understand, I am ok with supporting multi ports, in future we could use different port proxying TCP protocol

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 18 above)
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/network.go:357]: nit: use constants.RemoteGatewayClassName

Actual useful comment: I am a bit worried about tying this to the class.

We have a few use cases I think:
* I want to represent or expose a EW-GW as a service (today)
* I want to _represent_ a EW-GW as a Gateway (this PR)

But is there also a case where we want to actually _deploy_ a EW GW and use it as the cluster gateway? For example, make `class: aws-l4-external` or something? Would a user then have to make another `istio-remote` Gateway to access that? it feels a bit off

File: pilot/pkg/serviceregistry/kube/controller/network.go [hunk 21]
```diff
@@ -313,6 +333,86 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
 	return nil
 }
 
+func (n *networkManager) watchGatewayResources(c *Controller, stop <-chan struct{}) {
+	n.gatewayResourceClient = kclient.NewDelayedInformer(c.client, gvr.KubernetesGateway, kubetypes.StandardInformer, kubetypes.Filter{})
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/network.go:337]: nit: we can probably cast this to an `Informer[Gateway]` instead of untyped

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 16 above)
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:604]: Probably needs a feature flag

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 18 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/network.go:357]: Actually I see this is only for config cluster

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 19 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/network.go:390]: NamedAddress is deprecated

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/network.go:390]: Can ignore it and report in status

File: pilot/pkg/serviceregistry/kube/conversion.go [hunk 22]
```diff
@@ -233,3 +236,31 @@ func PodTLSMode(pod *corev1.Pod) string {
...
+	}
+	if l.Name == AutoPassthroughPortName && l.TLS.Mode != nil && *l.TLS.Mode == v1beta1.TLSModePassthrough {
+		return true
+	}
+	_, networkSet := gwLabels[label.TopologyNetwork.Name]
+	if !networkSet {
+		return false
+	}
+	expectedPort := "15443"
+	if port, f := gwLabels[label.NetworkingGatewayPort.Name]; f {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/conversion.go:262]: One note is this whole thing was kind of intended to be a temporary hack before we had it properly in upstream.. we may need to be careful about what is an is not experimental

File: tests/integration/pilot/workloadentry_test.go [hunk 23]
```diff
@@ -34,6 +34,27 @@ import (
 	"istio.io/istio/pkg/test/util/retry"
 )
 
+const ExistingDeploymentGateway = `
```
Comment (howardjohn) [inline tests/integration/pilot/workloadentry_test.go:37]: deadcode?

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 18 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:358]: This is when we specifically _don't_ want to deploy and if we don't want to deploy, we should only discover locally.  hope that makes sense.

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 16 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:623]: Yes, this is the only one now. Not quite sure if we should we expand it to gateway api.

File: pilot/pkg/serviceregistry/kube/controller/network.go (hunk 20 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/network.go:399]: Eventually we will probably start to support different types of port like HBONE

File: pilot/pkg/serviceregistry/kube/conversion.go (hunk 22 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/conversion.go:262]: > Before we had "it" properly in upstream

What is "it"?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/conversion.go:262]: `isAutoPassthrough`. Its really working around the lack of an explicit mode=auto passthrough

Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/conversion.go:262]: How would you feel about using options "istio.io/tlsMode: auto-passthrough" or some other custom option?

**Review by kdorosh — COMMENTED** [review]
is it intentional that `pilot/pkg/config/kube/gateway/testdata/eastwest-remote.yaml.golden` is empty?

**Comment by costinm** [comment]
I think long term auto passthrough will need to change - along with moving
to ServiceExport or equivalent explicit opt-in for multi cluster behavior.

And of course change to hbone with support for policies, etc.

It is not a future-prood feature - but short time may be ok for existing
non-ambient use.

File: pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml (hunk 5 above)
Comment (nmittler) [reply pilot/pkg/config/kube/gateway/testdata/eastwest-labelport.yaml:7]: > we can have multiple networks in one cluster.

Are you referring to the case of multiple nics?

**Comment by stevenctl** [comment]
> is it intentional that pilot/pkg/config/kube/gateway/testdata/eastwest-remote.yaml.golden is empty?

Yes. That gateway class should not generate config.

**Comment by stevenctl** [comment]
> I think long term auto passthrough will need to change - along with moving to ServiceExport or equivalent explicit opt-in for multi cluster behavior. And of course change to hbone with support for policies, etc. It is not a future-prood feature - but short time may be ok for existing non-ambient use.

Agree. I actually changed the way we detect auto passthrough to a more general definition where we can say some listener supports passthrough, bone or both.

File: manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml [hunk 24]
```diff
@@ -21,6 +21,9 @@ rules:
   - apiGroups: ["networking.istio.io"]
     verbs: [ "get", "watch", "list" ]
     resources: [ "workloadentries" ]
+  - apiGroups: ["networking.x-k8s.io", "gateway.networking.k8s.io"]
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml:24]: `networking.x-k8s.io` is gone, I don't think we need it anymore

File: manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml [hunk 25]
```diff
@@ -21,6 +21,9 @@ rules:
...
+    resources: ["gateways"]
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml:25]: This new dependency may be tricky. You mentioned when we don't have RBAC, everything is blocked? Is that right? Or at best it times out after 15s?

Do we have users always update this first before the control plane? I would think we don't?

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 16 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:624]: still needs a feature flag I think?

**Comment by kdorosh** [comment]
resolves https://github.com/istio/istio/issues/45172

File: manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml (hunk 25 above)
Comment (stevenctl) [reply manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml:25]: You'lll get 0 endpoints from that cluster. I have the feature flag off for now. 

I think if the flag is off by default for a few versions it should be ~safe enough. Skip version upgrades might have an issue though.

File: manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml (hunk 24 above)
Comment (stevenctl) [reply manifests/charts/istio-control/istio-discovery/templates/reader-clusterrole.yaml:24]: I was just copying what we had in the main ClusterRole for reading gatewayapi.

**Review by howardjohn — APPROVED** [review]

**Review by hzxuzhonghu — COMMENTED** [review]
We need to revist the behavior watching gateway from all clusters.

This is inconsistent with previous usage, like gateway deployment controller
